### ISSUE TITLE
Can user remap the hotkey in fugitive?

### ISSUE THREAD

**Issue by Freed-Wu** [tpope/vim-fugitive#1425]
Such as

```
n  s           &@:<C-U>execute <SNR>147_Do('Stage',0)<CR>
```

Because of `<SNR>`(`<SID>`), it is hard to remap. if use autoload function it can be easyer to remap.

Thanks!

**Comment by tpope** [comment]
I don't want to commit to these APIs publically. I am open to adding `<Plug>` maps but I want to do that thoughtfully, not just rush out the first thing that comes to mind.

My first question for this is always "Why?" Are you trying to support an alternative keyboard layout like Dvorak or Colemak? I have an undocumented interface for that.

**Comment by Freed-Wu** [comment]
I dare to think `map  xxx :call <SID>foobar()<CR>` is not better than `map xxx :call fugitive#foobar()<CR>` because it is hard for user to remap xxx. e.g. I `let g:mapleader='-'`, but fugitive `:<C-U>execute <SNR>189_Do('Toggle',1)<CR>` make me want to remap this toggle key to a localermap. And other user perhaps want to `:map <buffer> u <c-u>` or `:map  <buffer> d <c-d>`, but fugitive had map u and d to a `<SID>` map. It is really difficult to customize to map....
Thank you.

**Comment by tpope** [comment]
I do not want to make `s:Do()` public because it is a badly designed API, and making it public means I can no longer fix it because people are depending on it.

**Comment by Freed-Wu** [comment]
It is no need to rename or public... Can it provide a method for user to customize the map on the buffer fugitive? such as provide a dictionary like `g:fugitive_CommandMap={'-': '<localleader>t'}` for user to use `<localleader>t` to replace `-`?

**Comment by Figuera** [comment]
> 
> 
> I don't want to commit to these APIs publicly. I am open to adding `<Plug>` maps but I want to do that thoughtfully, not just rush out the first thing that comes to mind.
> 
> My first question for this is always "Why?". Are you trying to support an alternative keyboard layout like Dvorak or Colemak? I have an undocumented interface for that.

I do need to remap for a different keyboard layout. You mention a undocumented interface, could you give advice?

**Comment by phaazon** [comment]
Ping, I’m having the same problem (bépo keyboard, I keep having to rebind stuff when plugins provide default keybindings…).

**Comment by Figuera** [comment]
@phaazon,

I'm not a contributor to this project in any sense, but I did had to make the plugin compatible with my keyboard. Investigating the code I concluded that there is no user-friendly way of changing the mapping.

I ended up changing the source code myself, terrible solution since it will be overwritten everytime you upgrade the plugin. But at least is something. In `autoload/fugitive.vim` you will find most of the mapping being created. In my case I just wanted to stop the plugin from using the 'S' key therefore I changed the line 1971:

```
call s:Map('n', 's', ":<C-U>execute <SID>Do('Stage',0)<CR>", '<silent>')
```
to

```
call s:Map('n', 'j', ":<C-U>execute <SID>Do('Stage',0)<CR>", '<silent>')
```

that is I changed the `Stage` action from the key 'S' to the key 'J'. You could search for similar lines for the keys you need to change.

**Comment by phaazon** [comment]
Hell no, I won’t do that. Almost all plugins allow to disable default keybindings and provide a `keymap` map-like object. I’ll wait for @tpope to add such a thing, or I might make a PR if I find enough spare time. :)

**Comment by tpope** [comment]
The keyboard layout remap support is explained at https://github.com/tpope/vim-fugitive/issues/1080#issuecomment-521100430. It is a terrible fit for just about every other remapping use case, so I found it best to keep it out of the official documentation to avoid confusing users.

**Comment by vatosarmat** [comment]
@phaazon @Figuera how do you like this remapping method?
```vim
autocmd BufWinEnter */.git/index let @n="j\<cr>\<M-d>\<down>"
autocmd BufWinEnter */.git/index let @p="k\<cr>\<M-d>\<down>"
autocmd BufWinEnter */.git/index nnoremap <buffer><M-n> @n
autocmd BufWinEnter */.git/index nnoremap <buffer><M-p> @p
```
This allows to quickly overview delta by next-prev moving through changed files. <CR> selects file, `<M-d>` does `Gvdiffsplit! HEAD` and `<down>` moves focus back to fugitive summary window.

**Comment by 88maomao** [comment]
@vatosarmat thank you! I had to use your method to map `<tab>` to `=` in `fugitive`. btw. why `autocmd BufWinEnter */.git/index nnoremap <buffer><tab> =` doesn't work?
@tpope  My first attempt was `let g:nremap = {'=': '<TAB>'}`, but this led to broken mappings in [`tpope/vim-unimpaired`](https://github.com/tpope/vim-unimpaired/blob/4afbe5ebf32ad85341b4c02b0e1d8ca96a64c561/plugin/unimpaired.vim#L354) because it mapped `=p`  to `<tab>p` and `=P` to `<tab>P`.

**Comment by tpope** [comment]
`nnoremap` means "don't recursively call other maps", which is the opposite of what you want. Use `nmap`. Defining maps that call other maps like this is actually decent stopgap approach until this gets properly solved (which as you can probably guess, is buried under other priorities).

You should be able to use `autocmd FileType fugitive` for all these, rather than that weird-ass `BufWinEnter` event.

**Comment by ktchen14** [comment]
It seems like `autocmd FileType fugitive` isn't triggered on most buffers (e.g. `:Git log`).

**Comment by vatosarmat** [comment]
`GStatus` buffer definitely has `FileType fugitive`, but other buffers may have `git` filetype or something else

**Comment by tpope** [comment]
I was matching the weird-ass `BufWinEnter` event. For other buffers I would use `git` filetype plus check `&modifiable`, probably.

**Comment by ktchen14** [comment]
Hmm any idea how to do this for `:Git blame`? Seems like the line [here](https://github.com/tpope/vim-fugitive/blob/156dbcd738fbf997bfc996760d1f979a4496c746/autoload/fugitive.vim#L5799) is calling `s:BlameFileType()` to set the mappings. But because the `autocmd` is defined in an `autoload` file, it's hard to ensure that my

```
autocmd FileType fugitiveblame ...
```

Will be called *after* the one on that line.

**Comment by tpope** [comment]
I guess that can become an ftplugin.

**Comment by unphased** [comment]
I have unconventional normal mode maps for `-` and `=` which shrink and widen a vim window, the `-` clashes with `fugitiveblame` via the mapping `n  -           &@:<C-U>exe <SNR>102_BlameJump("")<CR> `, incidentally I can't work out what the "opposite" key to this is to undo this, so now I get stuck when I invariably use this muscle memory inside such windows. Hopefully soon it will become possible to remap or disable this.

**Comment by tpope** [comment]
Extracted ftplugin means you can `unmap <buffer> -` in `after/ftplugin/fugitiveblame.vim`.

**Comment by unphased** [comment]
It works! Thank you!

**Comment by weeman1337** [comment]
Hey, I have to questions:

> The keyboard layout remap support is explained at #1080 (comment).

I read the comment but I still don't know what to do. Some hotkeys don't work on my keyboard layout (DE).

The second question: How can I use for example F7 to navigate to the next hunk?

**Comment by tpope** [comment]
> Hey, I have two questions:
> 
> > The keyboard layout remap support is explained at #1080 (comment).
> 
> I read the comment but I still don't know what to do. Some hotkeys don't work on my keyboard layout (DE).

The dictionary contains Fugitive's maps as keys and your replacements as values. I've edited the comment to be a bit clearer.  See also [this deleted Unimpaired documentation](https://github.com/tpope/vim-unimpaired/commit/1a4bfc16b6254a8586f76eb976b9c57b7014c5e7) which was targeted at the DE keyboard problem of hard to type `[` and `]` keys.
 
> The second question: How can I use for example F7 to navigate to the next hunk?

I wouldn't use the dictionaries for that. Try:

    nmap <buffer> <F7> ]c

I'm assuming you want the `]c` version because `)` looks pretty easy to type on a DE keyboard.

**Comment by andrewrothman** [comment]
Hey everyone. I've been learning NeoVim. Really impressed with it and with the people in the (Neo)Vim community.

I've been enjoying using this remap while editing: `noremap <CR> :`

I noticed it didn't work in netrw, so I did this, which also lets me use the vanilla `<CR>` behavior by remapping it to `<Space>`.

```
augroup netrw_mapping
    autocmd!
    autocmd filetype netrw call NetrwMapping()
augroup END

function! NetrwMapping()
    nmap <buffer> <Space> <Plug>NetrwLocalBrowseCheck " previous behavior of <CR>
    nmap <buffer> <CR> :
endfunction
```

I'm just getting started understanding how to do these things. Is it easy for me to do the same with Fugitive?

**Comment by tpope** [comment]
`autocmd User FugitiveIndex,FugitiveTag,FugitiveCommit,FugitiveTree,FugitiveBlob` is an event that covers the status window and all historical buffer types. If you want to move the existing `<CR>` map to something else, you'll need to use `maparg()`. The full implementation is left as an exercise to the reader.

**Comment by cmonkey** [comment]
Hey, I have  questions:
        can I use for example C-P  to PreviouseItem 
  PS: hhkb keyborad (  Keys are more difficult to use

**Comment by tpope** [comment]
I continue to recommend mapping to the Fugitive map:

```
nmap <buffer> <C-N> )
nmap <buffer> <C-P> (
```

**Comment by rcywongaa-work** [comment]
For future vim noobs like me who had some trouble following, the full remapping command (e.g. to remap `-` to `gT`) would be
```
autocmd User FugitiveObject nmap <buffer> - gT
```

**Comment by GKreiken** [comment]
Hi everyone, noobie here trying to switch from VSC to VIM as my IDE. Note: I use colemak DH.

"issue" = I want to use "j" instead of "i" for the fugitive_i command.
After reading this discussion and the [comment](https://github.com/tpope/vim-fugitive/issues/1080#issuecomment-521100430), I was finally able to remap my "j" to "i" using
`autocmd User FugitiveObject,FugitiveIndex nmap <buffer> j i`

However, now both i and j do the same thing, as "i" is not unmapped from the command.

Using tpope's reccomended
```
let g:nremap = {'i': 'j'}
let g:xremap = {'i': 'j'}
let g:oremap = {'i': 'j'}
```
works, but this feels like using a bazooka for killing a musquito, because this sets some global vars right?

Sorry if this is a noobie question, thanks in advanced for any help!


P.S. vim, tpope, vim-community: you're awesome. Thank you!

**Comment by tpope** [comment]
This is the intended bazooka usage. They're global so that _all_ plugins that use `i` can use `j` instead.

**Comment by GKreiken** [comment]
Ok understood. I'll keep an eye out for unintended behaviour in other (already remapped) plugins. 

Thanks for the quick answer!

**Comment by tpope** [comment]
FWIW, I think the only plugins using this system are other plugins by me, and I don't think any of those map `i`.

**Comment by tpope** [comment]
It has dawned on me that I've overlooked some very low hanging fruit. If we toss out semantic naming and go with a literal scheme—`s` becomes `<Plug>fugitive:s`, etc.—then providing a `<Plug>` map for every existing map in the system becomes a small change to the helper method. I've just shipped the initial version.

I'm also providing `g:fugitive_defer_to_existing_maps` which does what's on the label. Between these 2 features, I think we can call this closed, even if it's a bit uglier than I'd like.

> It is no need to rename or public... Can it provide a method for user to customize the map on the buffer fugitive? such as provide a dictionary like `g:fugitive_CommandMap={'-': '<localleader>t'}` for user to use `<localleader>t` to replace `-`?

I've warmed up to this a bit since I first read it, and might consider tacking it on in the future. But let's start with my new experiment and see how well that meets people's needs.
